jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/internal/UpgradeRequestDelegate.java

```diff
@@ -73,14 +72,7 @@ public int getHeaderInt(String name)
     @Override
     public Map<String, List<String>> getHeaders()
     {
-        Map<String, List<String>> result = new LinkedHashMap<>();
-        HttpFields headers = request.getHeaders();
-        for (HttpField header : headers)
-        {
-            String name = header.getName();
-            result.put(name, headers.getValuesList(name));
-        }
-        return result;
+        return HttpFields.asMap(request.getHeaders());
```

Comment:
Sometimes you've made it immutable, other times you have not?

Reply:
I didn't look too deeply into this I just tried to maintain the behavior we had before. So if it was already immutable I just kept it that way.

But I will now do a deeper review of each case and assess whether it should be immutable or not.

Reply:
### Javax/Jakarta API
- On `JsrUpgradeListener.onHandshakeRequest`, the request headers should be mutable because you can set custom headers.
- During server websocket negotiation (`JsrHandshakeRequest`/`JsrHandshakeResponse`) only the request headers should be immutable because you can't change request headers.
- On the `JsrUpgradeListener.onHandshakeResponse`, the headers should be immutable because the response has already been received.

### Jetty API
- For the `ClientUpgradeRequest`, they should be mutable so request headers can be set.
- For the `DelegatedJettyClientUpgradeRequest` and `DelegatedJettyClientUpgradeResponse` they should be immutable because this is kept after the upgrade and aren't designed to be modified.
- For `UpgradeRequestDelegate` and `UpgradeResponseDelegate` they should be immutable because they are kept after the upgrade and are not intended to be modified.
- In `DelegatedServerUpgradeRequest` the headers should be immutable because this is during the negotiation.
- In `DelegatedServerUpgradeResponse` they should be mutable so response headers can be set in the upgrade response during negotiation.

Reply:
Just would like to point out that some people use the upgrade process to modify (strip out / remove) the websocket compression extension.

That way the actual upgrade does not use it and upgrades without compression.
I don't see a need for it to be immutable during upgrade.

But after upgrade, once the websocket is connected, the request is already done, nothing more is really used from it in a websocket sense.

Reply:
> some people use the upgrade process to modify (strip out / remove) the websocket compression extension

Yep, this is why the response headers are mutable during the server side negotiation. But it doesn't make sense for the request headers to be mutable at this stage.